text/0678-documenting-components.md

```diff
@@ -0,0 +1,326 @@
+}
+
+export default class ModalComponent extends Component<ModalArgs, ModalBlocks> {}
+```
+
+which in turn we can use in these two variants (with and without named blocks):
+
+```hbs
+<Modal as |close|>
+  Here is the content of the modal
+  
+  <button {{on "click" close}}>close</button>
+</Modal>
+
+<Modal @open={{true}}>
+  <:header as |close|>
+    This is my title
+    
+    <button {{on "click" close}}>close</button>  
+  </:header>
+  
+  <:body>
+    Here is the content of the modal
+  </:body>
+</Modal>
+```
+
+## Detailed design
+
+The design is sectioned into _definitions_, _documentation API_, _examples_ and then summarized with _benefits_.
+
+### Definitions
+
+At first let's define terms that will be used throughout this RFC:
+
+- **Block**
+  For any block invocation of a component, this describes the contents in there. We can distinguish between _default_ block and named blocks.
+  This is defined in syntax section for [RFC 460 - Yieldable Named Blocks](https://emberjs.github.io/rfcs/0460-yieldable-named-blocks.html#syntax)
+  
+  - _Default Block_
+    That is the default block when invoking a component:
+
+    ```hbs
+    <Modal>
+      this is the default block
+    </Modal>
+    ```
+    
+    equivalent to:
+    
+    ```hbs
+    <Modal>
+      <:default>this is the default block</:default>
+    </Modal>
+    ```
+
+  - _Named Block_
+    Named blocks can be referred to via an identifier and are represented through `<:identifier>` syntax:
+ 
+    ```hbs
+    <Modal>
+      <:header>Named header block</:header>
+      <:body>Named body block</:body>
+    </Modal>
+    ```
+- **Block Param**
+  Parameters passed from the component (through `{{yield}}`) to the consumer:
+  
+  ```hbs
+  <Modal as |close|>
+    <button {{on "click" close}}>close</button>
+  </Modal>
+  ```
+  
+  In the example `close` is a parameter _to_ the block. On the consuming side a space delimited list of parameters can be accessed within the pipe syntax (`| .. |`).
+  
+- **Yield**
+  Yielding is the mechanism from _inside_ the component to provide named blocks and pass parameters to them. Examples from RFC 460:
+  
+  ```hbs
+  <article>
+    <header>{{yield @article.title to='header'}}</header>
+    <section>{{yield @article.body to='body'}}</section>
+  </article>
+  ```
+  
+### Documentation API
+
+Documentation will use the powerful typing system of TypeScript. For TypeScript this works out of the box. For template-only components, use a `my-component.d.ts` file. For JavaScript, you can choose between a `*.d.ts` file or jsdoc with extended tsdoc annotations (see examples below).
+
+The interface for `@glimmer/component` needs extension for yields:
+
+```ts
+export default MyComponent extends Component<Input, Output> {}
+```
+
+The narrative is to give developers documentation slots for `Input` and `Output`. The `Input` will continue to work as is to document arguments and `Output` will be added to document blocks with their parameters.
```

Comment:
I'm not sure that `Input` and `Output` is the correct framing here. Blocks aren't exactly outputs, or at least, they are not the _sole_ outputs of the component. The entire template within the component is output.

Blocks are more like _callbacks_ in JavaScript. For instance:

```js
this.store.fetchRecord('foo', 123).then(() => {
  // do things...
});
```  

The callback passed to this promise is similar to a block. It gets passed as an argument, and internally it gets called later on at some point, and its result is _slotted_ into a specific place. This is basically what blocks do as well.

To me, blocks are just a different form of input, namespaced so that they cannot collide with arguments.

Reply:
This mental model preceeds blocks. Not accurate any longer. Agree, callbacks are more the mechanics that reflects this.